agent/dockerclient/dockerapi/docker_client.go

```diff
@@ -367,45 +382,79 @@ func wrapPullErrorAsNamedError(err error) apierrors.NamedError {
 	return retErr
 }
 
-func (dg *dockerGoClient) pullImage(image string, authData *apicontainer.RegistryAuthenticationData) apierrors.NamedError {
+func (dg *dockerGoClient) pullImage(ctx context.Context, image string, authData *apicontainer.RegistryAuthenticationData) apierrors.NamedError {
 	seelog.Debugf("DockerGoClient: pulling image: %s", image)
-	client, err := dg.dockerClient()
+	client, err := dg.sdkDockerClient()
 	if err != nil {
 		return CannotGetDockerClientError{version: dg.version, err: err}
 	}
 
 	sdkAuthConfig, err := dg.getAuthdata(image, authData)
-	authConfig := docker.AuthConfiguration{
-		Username:      sdkAuthConfig.Username,
-		Password:      sdkAuthConfig.Password,
-		Email:         sdkAuthConfig.Email,
-		ServerAddress: sdkAuthConfig.ServerAddress,
-	}
 	if err != nil {
 		return wrapPullErrorAsNamedError(err)
 	}
+	// encode auth data
+	var buf bytes.Buffer
+	if err := json.NewEncoder(&buf).Encode(sdkAuthConfig); err != nil {
+		return CannotPullECRContainerError{err}
+	}
 
-	pullDebugOut, pullWriter := io.Pipe()
-	defer pullWriter.Close()
+	imagePullOpts := types.ImagePullOptions{
+		All:          false,
+		RegistryAuth: base64.URLEncoding.EncodeToString(buf.Bytes()),
+	}
 
 	repository := getRepository(image)
 
-	opts := docker.PullImageOptions{
-		Repository:        repository,
-		OutputStream:      pullWriter,
-		InactivityTimeout: dg.config.ImagePullInactivityTimeout,
-	}
 	timeout := dg.time().After(dockerPullBeginTimeout)
 	// pullBegan is a channel indicating that we have seen at least one line of data on the 'OutputStream' above.
 	// It is here to guard against a bug wherein Docker never writes anything to that channel and hangs in pulling forever.
 	pullBegan := make(chan bool, 1)
-
-	go dg.filterPullDebugOutput(pullDebugOut, pullBegan, image)
+	// pullBeganOnce ensures we only indicate it began once (since our channel will only be read 0 or 1 times)
+	pullBeganOnce := sync.Once{}
 
 	pullFinished := make(chan error, 1)
-	// TODO Migrate Pull Image once Inactivity Timeout is sorted out
+	subCtx, cancelRequest := context.WithCancel(ctx)
+
 	go func() {
```

Comment:
Is this go routine along with the select statement necessary since we now implement the full go routine with the sdk change? It seems the select statement was only to catch a premature error (if nothing was being pulled), but now we can handle the error within the go routine without the select statement.

Reply:
given that images can be large and take long time to download, I think go routine here is necessary so that it is handled in a separate thread. I'm sure that this can be cleaned up, but I'd like to limit changes in behavior between go-dockerclient and SDK, and prefer to come back to this to make any improvements.

Reply:
We're still using the goroutine for the pullBegan timeout. You could rework that, but I think its fine to only focus on translating the API for now.